### PR TITLE
implemented Flatten Module

### PR DESCRIPTION
**Pull request by m0saan**
## Pull Request Description:
This pull request introduces a convenient way to flatten tensors in a flexible and readable manner.

**Changes Made:**
- Defined a `flatten` function that flattens a contiguous range of dimensions in a tensor.
- Created a `Flatten` module, a subclass of `Module`, which can be used to flatten input by reshaping it into a one-dimensional array.

**Details:**
- The `flatten` function takes an input tensor `a`, along with optional `start_dim` and `end_dim` parameters, and returns the flattened tensor.
- The `Flatten` module provides a high-level interface for flattening input tensors. It is initialized with optional `start_dim` and `end_dim` parameters.
- The module includes a documentation section with details about the input shape, output shape, and examples demonstrating the usage of the `Flatten` module.

**Examples:**
```python
import mlx.core as mx
import mlx.nn as nn

# Example 1: With default parameters
input_tensor = mx.random.normal((32, 3, 224, 224))
flattener = nn.Flatten()
output_tensor = flattener(input_tensor)
print(output_tensor.shape)  # Output: [32, 150528]

# Example 2: With non-default parameters
flattener = nn.Flatten(start_dim=0, end_dim=2)
output_tensor = flattener(input_tensor)
print(output_tensor.shape)  # Output: [21504, 224]
```

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/ml-explore/mlx/blob/main/CONTRIBUTING.md) document
- [x] I have run `pre-commit run --all-files` to format my code / installed pre-commit prior to committing changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated the necessary documentation (if needed)

### CODE REVIEW

File: python/mlx/nn/layers/flatten.py
Comment (awni) [inline python/mlx/nn/layers/flatten.py:46]: This should really be an op in `mx.core` (and follow the NumPy API please). Want to take a stab at moving it there and adding a Python binding? That would be awesome

**Review by awni — CHANGES_REQUESTED** [review]
Regarding the `flatten` op do you want to take a stab at moving it to core with a C++ op + binding?

File: python/mlx/nn/layers/flatten.py
Comment (m0saan) [reply python/mlx/nn/layers/flatten.py:46]: sure, I'm working on it! thanks for your input :p

Comment (m0saan) [reply python/mlx/nn/layers/flatten.py]: @awni can you please review up to this point?

File: mlx/ops.cpp
Comment (m0saan) [inline mlx/ops.cpp:259]: I've added a C++ impl of the flatten function :p

File: python/mlx/nn/layers/flatten.py
Comment (m0saan) [inline python/mlx/nn/layers/flatten.py:17]: Updated the python implementation to call `mx.flatten`

File: python/src/ops.cpp
Comment (m0saan) [inline python/src/ops.cpp:64]: I've added also the bindings for the C++ impl of flatten

File: python/mlx/nn/layers/flatten.py
Comment (awni) [reply python/mlx/nn/layers/flatten.py:17]: This function should be removed. We don't need both `nn.flatten` and `mx.flatten`.

File: mlx/ops.h
Comment (awni) [inline mlx/ops.h:122]: Following `numpy` we should not take a `start_dim` and `end_dim` in the `flatten` API. That should simplify the implementation a lot (you can get rid of the `getNewShape` stuff and just call `reshape` with a `{-1}`.

File: python/mlx/nn/layers/flatten.py
Comment (awni) [inline python/mlx/nn/layers/flatten.py:22]: I'm curious about your use case for this Module? Why not just do the reshape / flatten in the `__call__` function of your container Module? 

We are trying to avoid Module bloat (e.g. implementing a Module wrapper for every op in mlx.core) so to the extent that it's easy to just use the ops directly we prefer to go that route and keep the top-level `nn` clutter-free.

File: python/src/ops.cpp
Comment (awni) [inline python/src/ops.cpp:70]: Please remove these for compliance with numpy

**Review by awni — CHANGES_REQUESTED** [review]
Thanks for the updates!! Would be great to get the `mx.flatten` merged.  I left some comments, please address and let me know when it's ready for review.

File: python/mlx/nn/layers/flatten.py
Comment (m0saan) [reply python/mlx/nn/layers/flatten.py:22]: The use of nn.Flatten or similars like reshape in mlx is a matter of convenience and clarity in code rather than a strict necessity. The choice between using a specific module for flattening or doing the operation directly in the forward function depends on the specific requirements and design preferences.

for example we can use the flatten Module with the Sequential Module as follows:
```python
import mlx.core as mx
import mlx.nn as nn

sequential_net = nn.Sequential(
    nn.Conv2d(1, 32, kernel_size=3, stride=1, padding=1),
    nn.ReLU(),
    nn.Flatten(),
    nn.Linear(32 * 28 * 28, 10)  # Assuming input image size is 28x28
)

input_tensor = mx.random.normal((16, 1, 28, 28))  # Batch size of 1, 1 channel, 28x28 image

# Forward pass
output = sequential_net(input_tensor)
```

File: mlx/ops.h
Comment (m0saan) [reply mlx/ops.h:122]: Updating the flattening procedure through the exclusive use of `reshape(a, {-1} ...)`, consistently results in a 1D array. This differs from the behavior of the `flatten` operation. Consider a scenario involving a batch of images with the shape `[16, 3, 28, 28]`; the objective is to flatten the tensor from the second dimension onward, while preserving the batch dimension.

In a broader context:
The flattening process transforms the input by reshaping it into a one-dimensional tensor. If specific start_dim or end_dim parameters are provided, only dimensions ranging from start_dim to end_dim (inclusive) are flattened.

Comment (awni) [reply mlx/ops.h:122]: Ok, I think this is sufficiently useful that we can keep it.

File: python/mlx/nn/layers/flatten.py
Comment (awni) [reply python/mlx/nn/layers/flatten.py:22]: Yes, I certainly see the point. Nevertheless, I think it's a slippery slope to go this route as we have to start mirroring all of our ops with modules. I would prefer if you remove this for now and just land the `mx.flatten` op while we think about the right approach here.

For example, maybe something like a lambda module is a lot cleaner `nn.Lambda(fun)` which calls `fun` on it forward.

Comment (m0saan) [reply python/mlx/nn/layers/flatten.py:22]: Also the `nn.Flatten` module is designed to preserve the batch dimension while reshaping the input tensor into a one-dimensional format. It achieves this by collapsing all dimensions, except for the batch dimension. This eliminates the need to use `mx.flatten(a, start_dim=1)` for the same flattening operation. WDYT?

Comment (awni) [reply python/mlx/nn/layers/flatten.py:22]: Yea that’s a slight convenience but let’s hold off on adding it for now. Do you mind splitting it out of this PR? We will revisit the module decision after some discussion.

Comment (m0saan) [reply python/mlx/nn/layers/flatten.py:22]: Ahah, uure, I'll take care of removing the agreed-upon stuff and reach out to you.

**Comment by awni** [comment]
Do you mind if I push some changes to this? I added a few things in the C++  + some tests. I tried pushing them to your branch but I don't have permission. I think you can modify the permission in the PR to "allow edits from maintainers"

https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork

**Review by awni — APPROVED** [review]
Thanks a ton for this change!! Just FYI I added a few style changes and modified the C++ a bit.

**Pull request merged**
